src/modules/fw_pos_control_l1/FixedwingPositionControl.hpp

```diff
@@ -341,27 +345,59 @@ class FixedwingPositionControl final : public ModuleBase<FixedwingPositionContro
-	void		control_auto(const hrt_abstime &now, const Vector2d &curr_pos, const Vector2f &ground_speed,
-				     const position_setpoint_s &pos_sp_prev,
-				     const position_setpoint_s &pos_sp_curr, const position_setpoint_s &pos_sp_next);
+
+	/**
+	 * @brief Generate path tracking setpoints from waypoints
+	 *
+	 * @param now current time
+	 * @param curr_pos vehicle current position
+	 * @param ground_speed vehicle ground speed [m/s]
+	 * @param pos_sp_prev previous waypoint
+	 * @param pos_sp_curr current waypoint
+	 * @param pos_sp_next next waypoint
+	 * @return vehicle_local_path_setpoint_s path setpoints
+	 */
+	vehicle_local_path_setpoint_s pathHandler(const hrt_abstime &now, const Vector2d &curr_pos,
+			const Vector2f &ground_speed,
+			const position_setpoint_s &pos_sp_prev,
+			const position_setpoint_s &pos_sp_curr, const position_setpoint_s &pos_sp_next);
+
+	/**
+	 * @brief Get unit tangent of line segments between waypoints
+	 *
+	 * @param waypoint_A Current waypoint local position
+	 * @param waypoint_B Previous waypoint local position
+	 * @param vehicle_pos Current vehicle local position
+	 * @return Vector2f Unit tangent vector of the reference line segment
+	 */
+	Vector2f	navigateWaypoints(const Vector2f &waypoint_A, const Vector2f &waypoint_B, const Vector2f &vehicle_pos);
+
+	/**
+	 * @brief Fixedwing Auto position flight mode. Vehicle tracks waypoints / path
+	 *
+	 * @param now current time
+	 * @param curr_pos vehicle current local position
+	 * @param ground_speed vehicle ground speed
+	 * @param cruising_throttle cruising throttle of current waypoint
+	 * @param path_sp Reference path setpoint, pos_sp_prev, pos_sp_curr, pos_sp_next will be ignored if a path setpoint is provided
+	 */
+	void		control_auto(const hrt_abstime &now, const Vector2f &curr_pos, const Vector2f &ground_speed,
+				     const float &cruising_throttle, const vehicle_local_path_setpoint_s &path_sp);
 
 	void		control_auto_fixed_bank_alt_hold(const hrt_abstime &now);
 	void		control_auto_descend(const hrt_abstime &now);
 
-	void		control_auto_position(const hrt_abstime &now, const float dt, const Vector2d &curr_pos,
-					      const Vector2f &ground_speed,
-					      const position_setpoint_s &pos_sp_prev, const position_setpoint_s &pos_sp_curr);
-	void		control_auto_loiter(const hrt_abstime &now, const float dt, const Vector2d &curr_pos,
-					    const Vector2f &ground_speed,
-					    const position_setpoint_s &pos_sp_prev, const position_setpoint_s &pos_sp_curr, const position_setpoint_s &pos_sp_next);
-	void		control_auto_velocity(const hrt_abstime &now, const float dt, const Vector2d &curr_pos,
-					      const Vector2f &ground_speed,
-					      const position_setpoint_s &pos_sp_prev, const position_setpoint_s &pos_sp_curr);
-	void		control_auto_takeoff(const hrt_abstime &now, const float dt,  const Vector2d &curr_pos,
+	vehicle_local_path_setpoint_s		control_auto_position(const float dt, const Vector2f &curr_pos,
```

Comment:
so that I understand -- these two functions (also control_auto_loiter) are basically doing path handling for line segment and loiter that was previously done in NPFG / L1, resp.? The naming here is a bit confusing as there are a bunch of control_auto_XXX that exist in the state machine / mode manager function. maybe would be better to have path handler as separate library? that digests e.g. position setpoints, and converts to respective path setpoints? -- ideally, a path handler would take in path segment objects themselves.. but since we don't yet have this, and the legacy position setpoint triplet still needs to work, this could be a first iteration of path handler that just decodes the triplet into whatever path setpoint. In general - would be great to not have fw_pos_ctrl contain all of this logic.

Reply:
> maybe would be better to have path handler as separate library?

Yes, but currently the path handling itself relies in a lot of internal states. Therefore in order to separate it into a spearate library, we need to define the state flow without using internal state(which is not clear in the current state)

> but since we don't yet have this, and the legacy position setpoint triplet still needs to work, this could be a first iteration of path handler that just decodes the triplet into whatever path setpoint. In general - would be great to not have fw_pos_ctrl contain all of this logic.

I definitely agree. Therefore I think the next step would be to try to take the path handler out of fw_pos_control itself. However, I think the order should be that we clean up takeoff and landing before we do this.

Reply:
> maybe would be better to have path handler as separate library?

Yes, but currently the path handling itself relies in a lot of internal states. Therefore in order to separate it into a spearate library, we need to define the state flow without using internal state(which is not clear in the current state)

> but since we don't yet have this, and the legacy position setpoint triplet still needs to work, this could be a first iteration of path handler that just decodes the triplet into whatever path setpoint. In general - would be great to not have fw_pos_ctrl contain all of this logic.

I definitely agree. Therefore I think the next step would be to try to take the path handler out of fw_pos_control itself. However, I think the order should be that we clean up takeoff and landing before we do this.